### PR TITLE
Automatically push Docker image to Docker Hub

### PR DESCRIPTION
**Pull request by ErikSchierboom**
The [v3 development environment](https://github.com/exercism/development-environment/) uses Docker images from Docker Hub for all the tooling.
This allows for easy testing of the various tooling components (analyzers/representer/test-runners).

This PR adds a GitHub Actions workflow to automatically build and push a new Docker image whenever something is merged to master.

See https://github.com/exercism/v3/issues/2727

### CODE REVIEW

File: .github/workflows/dockerhub.yml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+jobs:
+  dockerhub:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+
+      - uses: satackey/action-docker-layer-caching@v0.0.5
```
Comment (SaschaMann) [inline .github/workflows/dockerhub.yml:15]:
```suggestion
      - uses: satackey/action-docker-layer-caching@da67a6cd88114ce3e232893dc6d946efa225aefb # 0.0.8
```

**Review by SaschaMann — APPROVED** [review]

**Review by SaschaMann — CHANGES_REQUESTED** [review]
Why do we have to build them twice? Can't we upload the image built in the deploy step instead?

**Comment by ErikSchierboom** [comment]
> Why do we have to build them twice? Can't we upload the image built in the deploy step instead?

We could, but it might be that in the future we want to do more in the deploy script than just building the image.

**Comment by SaschaMann** [comment]
> We could, but it might be that in the future we want to do more in the deploy script than just building the image.

We can still do that though. But building it twice wastes a large amount of resources for no good reason.

**Comment by iHiD** [comment]
There's a couple of reasons why we've got two scripts:
1) If DockerHub doesn't work for some reason (e.g. a timeout), we still want the deploy to happen (this scenario has already occurred to me)
2) I think it's quite likely the production and deployment builds will diverge. For example, you might want to have logging enabled in the dev (Dockerhub) one but not in production, or stripped dependencies in the production one. This is already case in the non-tooling repos (e.g. website).

If you want to merge them for now, then split them later, that's fine. Merge this then improve it. But in terms of us scripting everywhere, it's easier for us to keep them separate.

**Comment by SaschaMann** [comment]
> If DockerHub doesn't work for some reason (e.g. a timeout), we still want the deploy to happen (this scenario has already occurred to me)

That can be configured in the workflow.

> I think it's quite likely the production and deployment builds will diverge. For example, you might want to have logging enabled in the dev (Dockerhub) one but not in production, or stripped dependencies in the production one. This is already case in the non-tooling repos (e.g. website).

If that would happen, it could be dealt with once that arises.

> If you want to merge them for now, then split them later, that's fine. Merge this then improve it. But in terms of us scripting everywhere, it's easier for us to keep them separate.

I disagree with the approach "waste resources first, then fix it later". You can script-push a combined version everywhere and then fix it later when issues arise with that, too.

**Comment by iHiD** [comment]
So I've thought about this more and discussed with Erik. I'm happy with the two files approach for now. One is a file to help development, the other is a file for deployment. I understand that, in their current states, they could be optimised to be the same, but I don't think they're the same thing, and until we're x months into this, I feel like doing that everywhere would be premature optimisation. So I think I want the general approach to be that they're kept separate by default.

That said, I fully appreciate what you're saying and have no issue if you want to merge them for Julia, and manage that as things change over the coming months, but I don't think we should change the approach org-wide.

**Comment by SaschaMann** [comment]
> I understand that, in their current states, they could be optimised to be the same, but I don't think they're the same thing, and until we're x months into this, I feel like doing that everywhere would be premature optimisation.

It's not premature optimisation, it's the other way around. Now you're accounting for issues that may or may not even occur that could easily be resolved later on.

> but I don't think they're the same thing

They are. They build the same image but then do slightly different things. You're speculating that at some point in the future they may diverge but as of right now they're the same thing (with different names at most).

**Comment by iHiD** [comment]
I disagree :) They're different concerns and I want to treat them differently.

**Review by SaschaMann — APPROVED** [review]

**Pull request merged**